### PR TITLE
Persistent events queue full

### PR DESCRIPTION
**Pull request by edengal**

### CODE REVIEW

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java [hunk 1]
```diff
@@ -28,6 +26,9 @@
 
 public class ElasticsearchUtil {
 	public static final String CLIENT = "client";
+	public static final String DISK_HANDLER = "disk_handler";
+	public static final String EVENTS_QUEUE = "events_queue";
+	public static final String OVERFLOWED_EVENTS_QUEUE = "overflowed_events_queue";
 	public static final int THREAD_SLEEP = 2000;
 	public static final String SCRIPT =
```
Comment (DvirKatz) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java:33]: what do you say about moving it to resources dir and read from there?

Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java:33]: why?

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java [hunk 2]
```diff
@@ -81,9 +90,64 @@ private void init(){
...
+	}
+
+	private List<Event> fetchAndDeleteOverflowedEvents(boolean deleteAfterFetch) {
+		List<Event> allEvents = new ArrayList<>();
+		ISqlJetCursor resultCursor = null;
+		int fetchedCount = 0;
+		List<Event> events;
+
+		try {
+			db.beginTransaction(SqlJetTransactionMode.WRITE);
```
Comment (DvirKatz) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java:104]: shouldn't this be READ?

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java [hunk 3]
```diff
@@ -81,9 +90,64 @@ private void init(){
...
+			resultCursor = overFlowedEventsTable.lookup(overFlowedEventsTable.getPrimaryKeyIndexName());
+
+			while (fetchedCount < maxFetchedBulksInOneTime && !resultCursor.eof()) {
```
Comment (DvirKatz) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java:107]: i wonder if we need to fetch from db with the cursor, I mean we already fetching not more than _maxFetchedBulksInOneTime_ so there is no risk of getting a lot of events and killing the db.
or is that for the deletion using the cursor?

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/EventsPersistentFetchJob.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+			while (hasEnoughRoomLeft(eventsQueue)){
+				List<Event> events = diskHandler.fetchAndDeleteOverflowedEvents();
+				if (events.isEmpty()){
+					break;
+				}
+				else {
+					for (Event event : events) {
+						if (!eventsQueue.offer(event)) {
+							if (!overflowedQueue.offer(event)) {
+								LOG.error("OverflowedQueue is full, event {} was discarded", event.getTaskId());
```
Comment (DvirKatz) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/EventsPersistentFetchJob.java:37]: don't we want to return it to DB?

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java (hunk 1 above)
Comment (DvirKatz) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java:33]: it makes this class huge and ugly

Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/common/ElasticsearchUtil.java:33]: I kinda like it here. Without the hassle of reading a file, writing a file.

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java (hunk 2 above)
Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java:104]: right

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java (hunk 3 above)
Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/SQLJetDiskHandler.java:107]: a bit easier for deletion. can do without the cursor and delete in a query but no real difference.

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/EventsPersistentFetchJob.java (hunk 4 above)
Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/EventsPersistentFetchJob.java:37]: @DvirKatz We are, by pushing it back to the overflowedQueue
If it's full we discard, same as regular full queue.

**Pull request merged**